src/line_parser.rs

```diff
@@ -0,0 +1,299 @@
+use wmidi::Note;
+
+use crate::{BeatNumber, Line, LineNote};
+
+#[derive(Debug)]
+enum State {
+    AwaitingNextToken,
+    ParsedNoteLetter { note_letter: char },
+    ParsedNoteName { note_name: String },
+}
+
+pub fn parse_line(line_str: &str) -> Line {
+    LineParser::new(line_str).parse()
+}
+
+struct LineParser {
+    chars: Vec<char>,
+    next_index: usize,
+    state: State,
+    currently_sustained_note: Option<(Note, BeatNumber)>,
+    current_beat_number: BeatNumber,
+}
+
+impl LineParser {
+    pub fn new(line_str: &str) -> Self {
+        Self {
+            chars: line_str.chars().collect(),
+            next_index: 0,
+            state: State::AwaitingNextToken,
+            currently_sustained_note: None,
+            current_beat_number: BeatNumber { sixteenth_note: 0 },
+        }
+    }
+
+    pub fn parse(&mut self) -> Line {
+        let mut line_notes: Vec<LineNote> = vec![];
+
+        while self.next_index < self.chars.len() {
+            let next_char = self.chars[self.next_index];
+            match (next_char, &self.state) {
+                (' ', State::AwaitingNextToken) => {
+                    self.forward(1);
+                }
+                ('C' | 'D' | 'E' | 'F' | 'G' | 'A' | 'B', State::AwaitingNextToken) => {
+                    self.finish_note_if_sustaining(&mut line_notes);
+                    self.state = State::ParsedNoteLetter {
+                        note_letter: next_char,
+                    };
+                    self.forward(1);
+                }
+                ('b', State::ParsedNoteLetter { note_letter }) => {
+                    self.state = State::ParsedNoteName {
+                        note_name: vec![note_letter, &next_char].into_iter().collect(),
+                    };
+                    self.forward(1);
+                }
+                (
+                    // TODO: handle -1
+                    '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9',
+                    State::ParsedNoteLetter { note_letter },
+                ) => {
+                    let note_letter = note_letter.to_string();
+                    self.mark_sustaining_note(note_letter, next_char);
+                    self.finished_parsing_beat();
+                    self.forward(1);
+                }
+                (
+                    '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9',
+                    State::ParsedNoteName { note_name },
+                ) => {
+                    let note_name = note_name.to_string();
```

Comment:
This is some extra string-cloning shenanigans to avoid a borrowing issue where `.mark_sustaining_note()` takes a mutable reference to `self` but then the compiler was unhappy that I was simultaneously passing in a reference derived from an immutable borrow of `self` (namely `&self.state` in the enclosing `match` here, I think)